k6 & Grafana 부하테스트 및 상품등록 기능 구현(판매자 모듈)

KWAK-JINHO opened the pull request:
### 🚀 어떤 기능을 구현했나요 ?
- 판매자 모듈 생성
- 상품(Product), 판매상품(SaleProduct), 재고(Stock) 관련 도메인 모델 설계
- 상품 등록 기능 구현 및 성능 최적화
- k6 + grafana + influxDB 조합으로 상품 목록조회 기능 부하테스트 진행

### 🔥 어떤 문제를 마주했나요 ?
- 대량의 상품 옵션 동시 등록 시 성능 저하 문제

### ✨ 어떻게 해결했나요 ?
- 일괄 처리 로직 구현으로 N+1 쿼리 문제 해결 (saveAll 메소드 활용)

### 📝 어떤 부분에 집중해서 리뷰해야 할까요?
- SaleProductServiceImpl의 registerSaleProduct 메소드에서 옵션과 재고를 동시에 처리하는 로직
- 상품 등록 시 여러 엔티티 간의 트랜잭션 처리 방식
- SaleProduct와 Stock 객체를 효율적으로 저장하는 방식

KWAK-JINHO committed:
🐛 fix: sort 파라미터가 null일 때 NullPointerException 오류 수정

- sort 값이 null로 입력될 경우 @Pattern 검증에서 발생하던 NullPointerException 해결
- 컨트롤러 레벨에서 null 값 사전 처리 로직 추가

🔗 Resolves: #318

KWAK-JINHO committed:
✨ feat: k6를 이용한 상품 검색 API 부하테스트 환경 구축

- k6, influxDB, Grafana를 Docker Compose로 통합 구성
- 다양한 검색 파라미터를 포함한 상품 API 테스트 스크립트 작성
- 동시 사용자 1,000명, 1분 동안의 부하 시나리오 구현

🔗 Resolves: #318

KWAK-JINHO committed:
🔧 chore: 코드 컨벤션 수정

🔗 Resolves: #318

KWAK-JINHO committed:
🛠️ refactor: Page 응답 공통 response를 사용하게 수정

- page객체를 두번 생성하던 구조에서 공통 객체를 만들어 한번만 생성하게 변경

🔗 Resolves: #318

KWAK-JINHO committed:
🛠️ refactor: Page 응답 공통 response를 사용하게 수정2

🔗 Resolves: #318

KWAK-JINHO committed:
🛠️ refactor: Product Label nullable = false로 수정

🔗 Resolves: #318

KWAK-JINHO committed:
🛠️ refactor: mallSeq Type Long으로 수정

🔗 Resolves: #318

KWAK-JINHO committed:
🛠️ refactor: Product 와 saleProducts 연관관계수정

- Product 삭제시 saleProducts 삭제되게 변경

🔗 Resolves: #318

KWAK-JINHO committed:
🔧 chore: SizeOption 메소드에서 필드 접근 방식 개선

- record 내부에서 게터 메소드 대신 직접 필드 접근 방식으로 수정하여 코드 간결성 향상

🔗 Resolves: #318

KWAK-JINHO committed:
🔧 chore: Status 생성자 매개변수 순서 병경

🔗 Resolves: #318

KWAK-JINHO committed:
🔧 chore: 더이상 사용하지 않는 메서드 삭제

🔗 Resolves: #318

KWAK-JINHO committed:
🛠️ refactor: getMallSeq 타입 변경에 따른 수정

🔗 Resolves: #318

KWAK-JINHO committed:
🛠️ refactor: Product 엔티티 생성자에서 Category 엔티티 의존성 제거

🔗 Resolves: #318

KWAK-JINHO committed:
✨ feat: 판매자 애플리케이션 모듈 분리

🔗 Resolves: #318

KWAK-JINHO committed:
✨ feat: (판매자 모듈) 글로벌 에러처리 기능

🔗 Resolves: #318

KWAK-JINHO committed:
✨ feat: (판매자 모듈) 카테고리 기능

🔗 Resolves: #318

KWAK-JINHO committed:
✨ feat: (판매자 모듈) 상품 옵션 기능

🔗 Resolves: #318

KWAK-JINHO committed:
✨ feat: (판매자 모듈) ProductScore 기능

🔗 Resolves: #318

KWAK-JINHO committed:
✨ feat: (판매자 모듈) 상품 등록 기능 구현

🔗 Resolves: #318

KWAK-JINHO committed:
✨ feat: (판매자 모듈) SaleProduct 도메인

🔗 Resolves: #318

KWAK-JINHO committed:
✨ feat: (판매자 모듈) Stock 도메인

🔗 Resolves: #318

KWAK-JINHO committed:
✨ feat: (판매자 모듈) Store 도메인

🔗 Resolves: #318

KWAK-JINHO committed:
🛠️refactor: saleProduct 사라지면 Stock도 사라지게 수정

🔗 Resolves: #318

KWAK-JINHO committed:
🔧chore: (판매자 모듈) 판매상품등록 dto 이름수정

🔗 Resolves: #318

KWAK-JINHO committed:
✨ feat: 상품 등록과 판매상품 등록 기능 구현

- 옵션별 판매상품 자동 생성 로직 구현
- 상품 옵션 존재 여부 검증 로직 개선

🔗 Resolves: #318

KWAK-JINHO committed:
✨ feat: 재고 등록 기능 구현

- 판매상품 등록시 해당 재고 등록

🔗 Resolves: #318

KWAK-JINHO committed:
🐛 fix: 요청값 optionIds 유효성 검증 수정

- NotEmpty로 리스트가 비어있는지, <@NotNull Long>를 통해 요소들중에 null이 없는지 체크

🔗 Resolves: #318

KWAK-JINHO committed:
🛠️ refactor: SaleProduct와 Stock 저장 로직 최적화

- Stock 등록을 위한 서비스 메서드 리팩토링
- 개별 save() 대신 saveAll() 메서드로 일괄 저장 구현

🔗 Resolves: #318

ehouse16 commented on backend/JiSeller/src/main/java/com/jiseller/option/domain/Option.java:
```diff
@@ -0,0 +1,26 @@
+public class Option extends BaseEntity {
+
+    @Enumerated(EnumType.STRING)
+    private OptionCategory categoryType;
```
`nullable`처리를 안해도 괜찮은가요?

KWAK-JINHO replied:
nullable 추가하였습니다! 감사합니당

ehouse16 commented on backend/JiSeller/src/main/java/com/jiseller/saleproduct/domain/SaleProduct.java:
여기도 `nullable` 처리가 있으면 좋을 거 같아요!

KWAK-JINHO replied:
감사합니다~

ehouse16 commented on backend/JiSeller/src/main/java/com/jiseller/stock/dto/RegisterStockRequest.java:
```diff
@@ -0,0 +1,13 @@
+import com.jiseller.stock.domain.Stock;
+
+public record RegisterStockRequest(
+        int quantity,
```
`request`니까 `validation`을 추가하는 건 어떻게 생각하시나요?

+ 쭉 읽어보니까 controller 가 없네요? 이럴때는 validation이 있는 게 좋을까요 아니면 없어도 무관할까요? (순수 궁금증입니다,,)

KWAK-JINHO replied:
해당 도메인은 사용자와 직접 소통하지 않고, 상품등록 과정에서 선택된 옵션과 연계하여 판매상품을 등록합니다. 이때 1:1 관계인 Stock이 등록되므로 별도의 컨트롤러는 구현하지 않았습니다.
저도 validation을 어느 수준까지 구현해야 하는지에 대해 고민이 많습니다. 향후 유지보수를 고려했을 때 비용을 생각하지 않으면 데이터 전달이 발생하는 모든 지점에서 검증 처리를 하는 것이 더 견고한 방식이 될 수 있다고는 생각합니다.

KWAK-JINHO commented on backend/JiSeller/src/main/java/com/jiseller/product/controller/ProductController.java:
아직 프론트와 협업 계획이 없는 모듈이라 아직 스웨거를 이용한 API 명세가 필요없다 판단해 따로 Impl을 만들지 않았습니다.

KWAK-JINHO committed:
🛠️ refactor: 엔티티 클래스의 필수 필드에 `nullable = false` 속성 추가

🔗 Resolves: #318

KWAK-JINHO commented on backend/JiSeller/src/main/java/com/jiseller/product/domain/embed/Status.java:
```diff
@@ -0,0 +1,44 @@
+@Getter
+@Embeddable
+@NoArgsConstructor(access = AccessLevel.PROTECTED)
+public class Status {
```
판매자 모듈은 추후 구매자 사이트와 분리되어 별도의 서비스가 될것이기 때문에 클래스명을 일일이 구별하지 않았습니다.

KWAK-JINHO commented on backend/JiSeller/src/main/java/com/jiseller/product/service/ProductServiceImpl.java:
```diff
@@ -0,0 +1,37 @@
+    public void registerProduct(final RegisterProductRequest registerProductRequest) {
+        categoryRepository.findById(registerProductRequest.lCatId())
+                .orElseThrow(() -> new DomainException(ErrorType.CATEGORY_NOT_FOUND));
+        final Store store = storeRepository.findByMallSeq(registerProductRequest.mallSeq())
```
구매자 사이트와 동일한 이유입니다. final 을 붙임으로써 명시적으로 변경되지 않을것이란걸 알 수 있으며, 의도치 않은 변수의 값 변경을 방지할 수 있다고 생각했습니다. 또한 final 키워드가 가비지 컬렉션을 향상시킨다는 말이 있었는데 이는 사실이 아닌것 같습니다.
https://stackoverflow.com/questions/306862/does-using-final-for-variables-in-java-improve-garbage-collection
링크 참고해보시면 도움될것 같습니다.

KWAK-JINHO commented on backend/JiSeller/src/main/java/com/jiseller/product/dto/RegisterProductRequest.java:
```diff
@@ -0,0 +1,112 @@
+        return new CategoryInfo(lCatId, mCatId, sCatId);
+    }
+
+    private ImageUrl createImageUrl() {
```
당장 객체를 만들때 필요한 값이라 하드코딩해둔 상태입니다.

KWAK-JINHO committed:
🛠️ refactor: util 디렉토리를 생성하여 DTO와 책임분리

- mapper와 dto 그리고 책임을 명시하기 위한 래퍼 클래스 분리

🔗 Resolves: #318

KWAK-JINHO committed:
# 커밋 유형:
# ✨ feat: (새 기능) | 🐛 fix: (버그 수정) | 🛠️ refactor: (리팩토링) |
# 📝 docs: (문서) | ♻️ test: (테스트) | 🔧 chore: (기타 작업)
#
# 커밋 메세지 제목: 구현 사항을 한 줄로 요약해 간략하게 작성
# 커밋 메세지 본문: 무엇을, 어떻게 구현했는지 설명 (가독성을 고려한 줄바꿈 필수!)
# 🔗 꼬리말: 어떤 브랜치에서 작성한 커밋인지 작성 브랜치 번호를 작성

🛠️ refactor: 예외 처리 핸들러 리펙토링

- 바인딩 결과에서 필드명과 오류 메시지 수집하여 디버깅에 용이하게 수정

🔗 Resolves: #318

KWAK-JINHO committed:
🐛 fix: 의존성 수정

🔗 Resolves: #318

KWAK-JINHO committed:
🐛 fix: DB 의존성 수정

🔗 Resolves: #318

KWAK-JINHO committed:
🐛 fix: 테이블명 명시 (판매자 모듈)

- DB 공유에 따라 엔티티명 명시

🔗 Resolves: #318

KWAK-JINHO committed:
🐛 fix: 숫자 필드 유효성 검사를 @NotBlank에서 @NotNull로 수정

- 숫자 타입 필드에 @NotBlank 대신 @NotNull 어노테이션 적용

🔗 Resolves: #318

KWAK-JINHO committed:
🔧 chore: 판매자/구매자 모듈 공용 DB 설정 추가

- 판매자 모듈과 구매자 모듈이 동일한 ShopDB를 사용하도록 JPA 및 DataSource 설정 적용

🔗 Resolves: #318

KWAK-JINHO committed:
🛠️ refactor: 카테고리 조회 메서드 findBytId → findByCurrentId로 변경

- 상품 등록 시 카테고리 조회 로직에서 findBytId 대신 findByCurrentId 메서드 사용

🔗 Resolves: #318

KWAK-JINHO committed:
✨ feat: 옵션 등록 기능 추가

- 옵션 컨트롤러, 서비스, 레포지토리 추가
- 옵션 등록 요청 DTO 추가

🔗 Resolves: #318

KWAK-JINHO committed:
🛠️ refactor: 옵션 등록 시 유효성 검증 추가

- FASHION_CLOTHES 타입에 대한 옵션값 형식 검증 로직을 DTO로 이동

🔗 Resolves: #318
